packages/@aws-cdk/aws-iam/lib/policy-statement.ts

```diff
@@ -408,6 +409,27 @@ export class PolicyStatement {
     }
     this.addConditions(conditions);
   }
+
+  /**
+   * Validate PolicyStatementProps
+   *
+   * As per [IAM policy grammar](https://docs.aws.amazon.com/en_us/IAM/latest/UserGuide/reference_policies_grammar.html#policies-grammar-bnf),
+   * the action and resource blocks are mandatory.
+   *
+   * @param props
+   */
+  private validateProps(props: PolicyStatementProps) {
+    if (props.conditions || props.effect || props.notPrincipals || props.notResources || props.principals || props.resources || props.sid) {
```

Comment:
I don't quite get this additional guarding `if`. Is `action` always required, or not? Where are you running into trouble if you don't add this?

Reply:
The additional `if` is to guard against `PolicyStatement` class instantiated without any props.
```ts
const p = new PolicyStatement();
```

Reply:
Ah the old usage pattern. Good catch. Could you rewrite it a little though to make that more self-evident for the future?

Maybe to the following:

```ts
const anyConstructorParamsSupplied = Object.values(props).filter(x => x !== undefined).length > 0;

if (anyConstructorParamsSupplied && !(props.actions || props.notActions)) {
  throw new Error('...');
}
```

Reply:
Hey but while writing that, that reminds me...

We can *force* either a full policy statement in the constructor, or an empty policy statement in the constructor and then using the fluent API... but why would we prevent a mix?

I think it's actually a breaking change to do that.

People could have written:

```ts
const s = new PolicyStatement({ resources: ['*'] });
s.addPrincipal(...):
```

And this change would break them.

The only place where we can do proper validation is at the end, when we are rendering the policy statement out.